pkg/alertmanager/multitenant.go

```diff
@@ -571,51 +570,49 @@ func (am *MultitenantAlertmanager) stopping(_ error) error {
 
 // loadAlertmanagerConfigs Loads (and filters) the alertmanagers configuration from object storage, taking into consideration the sharding strategy.
 func (am *MultitenantAlertmanager) loadAlertmanagerConfigs(ctx context.Context) (map[string]alertspb.AlertConfigDesc, error) {
-	configs, err := am.store.ListAlertConfigs(ctx)
+	// Find all users with an alertmanager config.
+	userIDs, err := am.store.ListAllUsers(ctx)
 	if err != nil {
-		return nil, err
-	}
-
-	// Without any sharding, we return _all_ the configs and there's nothing else for us to do.
-	if !am.cfg.ShardingEnabled {
-		am.tenantsDiscovered.Set(float64(len(configs)))
-		am.tenantsOwned.Set(float64(len(configs)))
-		return configs, nil
+		return nil, errors.Wrap(err, "failed to list users with alertmanager configuration")
 	}
+	numUsersDiscovered := len(userIDs)
 
-	ownedConfigs := map[string]alertspb.AlertConfigDesc{}
-	for userID, cfg := range configs {
-		owned, err := am.isConfigOwned(userID)
-		if err != nil {
-			am.ringCheckErrors.Inc()
-			level.Error(am.logger).Log("msg", "failed to load alertmanager configuration for user", "user", userID, "err", err)
+	// Filter out users not owned by this shard.
+	for i := 0; i < len(userIDs); {
+		if !am.isUserOwned(userIDs[i]) {
+			userIDs = append(userIDs[:i], userIDs[i+1:]...)
 			continue
 		}
 
-		if owned {
-			level.Debug(am.logger).Log("msg", "alertmanager configuration owned", "user", userID)
-			ownedConfigs[userID] = cfg
-		} else {
-			level.Debug(am.logger).Log("msg", "alertmanager configuration not owned, ignoring", "user", userID)
-		}
+		i++
 	}
+	numUsersOwned := len(userIDs)
 
-	am.tenantsDiscovered.Set(float64(len(configs)))
-	am.tenantsOwned.Set(float64(len(ownedConfigs)))
-	return ownedConfigs, nil
+	// Load the configs for the owned users.
+	configs, err := am.store.GetAlertConfigs(ctx, userIDs)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to load alertmanager configurations for owned users")
+	}
+
+	am.tenantsDiscovered.Set(float64(numUsersDiscovered))
+	am.tenantsOwned.Set(float64(numUsersOwned))
+	return configs, nil
 }
 
-func (am *MultitenantAlertmanager) isConfigOwned(userID string) (bool, error) {
-	ringHasher := fnv.New32a()
-	// Hasher never returns err.
-	_, _ = ringHasher.Write([]byte(userID))
+func (am *MultitenantAlertmanager) isUserOwned(userID string) bool {
```

Comment:
it's easier to use if it can't return `error` and the error case is handled internally.